File: examples/12-signals-and-safe-handling/go/main.go
// pmon v1 — a mini process supervisor (chapter 12: signals and safe handling).
//
//	pmon run -- CMD [ARGS...]
//	pmon supervise [--max-restarts N] [--backoff-ms B] -- CMD [ARGS...]
//
// The Go take on async-signal-safe design: the runtime already owns the
// signal handlers (they do nothing but forward a note to the runtime), and
// os/signal turns delivery into ordinary channel sends. Supervision is then
// a single select over three channels: signals from signal.Notify, child
// exits from a wait goroutine, and a backoff timer. No user code ever runs
// in signal context.
package main

import (
	"fmt"
	"os"
	"os/exec"
	"os/signal"
	"strconv"
	"syscall"
	"time"

	"golang.org/x/sys/unix"
)

const usageText = `usage: pmon run -- CMD [ARGS...]
       pmon supervise [--max-restarts N] [--backoff-ms B] -- CMD [ARGS...]`

func usage() {
	fmt.Fprintln(os.Stderr, usageText)
	os.Exit(2)
}

func die(err error) {
	fmt.Fprintf(os.Stderr, "pmon: error: %v\n", err)
	os.Exit(1)
}

// childExit is the decoded wait status of a reaped child.
type childExit struct {
	status   int
	signo    int
	signaled bool
}

func decode(ps *os.ProcessState) childExit {
	ws, ok := ps.Sys().(syscall.WaitStatus)
	if ok && ws.Signaled() {
		return childExit{signo: int(ws.Signal()), signaled: true}
	}
	return childExit{status: ps.ExitCode()}
}

func report(pid int, ce childExit) {
	if ce.signaled {
		fmt.Printf("pmon: child %d killed signal=%d\n", pid, ce.signo)
	} else {
		fmt.Printf("pmon: child %d exited status=%d\n", pid, ce.status)
	}
}

// spawn starts CMD with inherited stdio and prints the started line.
func spawn(argv []string) (*exec.Cmd, error) {
	cmd := exec.Command(argv[0], argv[1:]...)
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("start %s: %w", argv[0], err)
	}
	fmt.Printf("pmon: started pid %d\n", cmd.Process.Pid)
	return cmd, nil
}

func runOnce(argv []string) int {
	cmd, err := spawn(argv)
	if err != nil {
		die(err)
	}
	_ = cmd.Wait() // error carries the nonzero status; ProcessState has it all
	ce := decode(cmd.ProcessState)
	report(cmd.Process.Pid, ce)
	if ce.signaled {
		return 128 + ce.signo
	}
	return ce.status
}

func signame(s os.Signal) string {
	switch s {
	case unix.SIGTERM:
		return "SIGTERM"
	case unix.SIGINT:
		return "SIGINT"
	default:
		return s.String()
	}
}

type superviseOpts struct {
	maxRestarts int
	backoffMs   int64
	argv        []string
}

func supervise(opts superviseOpts) int {
	sigCh := make(chan os.Signal, 8)
	signal.Notify(sigCh, unix.SIGTERM, unix.SIGINT, unix.SIGHUP)

	// One in-flight wait goroutine per live child; the buffered channel lets
	// it finish even when the supervisor consumes the exit via stopChild.
	waitCh := make(chan childExit, 1)
	start := func() *exec.Cmd {
		cmd, err := spawn(opts.argv)
		if err != nil {
			die(err)
		}
		go func() {
			_ = cmd.Wait()
			waitCh <- decode(cmd.ProcessState)
		}()
		return cmd
	}
	// SIGTERM the child and consume its exit. Used by shutdown and reload.
	stopChild := func(cmd *exec.Cmd) {
		_ = unix.Kill(cmd.Process.Pid, unix.SIGTERM) // ESRCH: already exiting
		<-waitCh
	}

	cmd := start()
	restarts := 0
	backoff := opts.backoffMs
	var timer *time.Timer // non-nil: no child alive, waiting out a backoff

	for {
		var timerCh <-chan time.Time
		if timer != nil {
			timerCh = timer.C
		}
		select {
		case s := <-sigCh:
			switch s {
			case unix.SIGTERM, unix.SIGINT:
				if timer == nil {
					stopChild(cmd)
				}
				fmt.Printf("pmon: shutting down (%s)\n", signame(s))
				return 0
			case unix.SIGHUP:
				fmt.Println("pmon: reload requested")
				if timer == nil {
					stopChild(cmd)
				} else {
					timer.Stop()
					timer = nil
				}
				restarts = 0
				backoff = opts.backoffMs
				cmd = start()
			}
		case ce := <-waitCh:
			// The child exited on its own.
			report(cmd.Process.Pid, ce)
			if !ce.signaled && ce.status == 0 {
				return 0
			}
			if restarts >= opts.maxRestarts {
				fmt.Printf("pmon: giving up after %d restarts\n", opts.maxRestarts)
				return 1
			}
			restarts++
			fmt.Printf("pmon: restart #%d (backoff %dms)\n", restarts, backoff)
			timer = time.NewTimer(time.Duration(backoff) * time.Millisecond)
			backoff *= 2
		case <-timerCh:
			// Backoff elapsed: bring the child back.
			timer = nil
			cmd = start()
		}
	}
}

func parseSupervise(args []string) (superviseOpts, error) {
	opts := superviseOpts{maxRestarts: 5, backoffMs: 100}
	i := 0
	numericFlag := func(name string, min int64) (int64, error) {
		if i+1 >= len(args) {
			return 0, fmt.Errorf("%s needs a value", name)
		}
		i++
		v, err := strconv.ParseInt(args[i], 10, 64)
		if err != nil || v < min {
			return 0, fmt.Errorf("bad value for %s", name)
		}
		return v, nil
	}
	for i < len(args) {
		switch args[i] {
		case "--":
			i++
			opts.argv = args[i:]
			if len(opts.argv) == 0 {
				return opts, fmt.Errorf("no command after --")
			}
			return opts, nil
		case "--max-restarts":
			v, err := numericFlag("--max-restarts", 0)
			if err != nil {
				return opts, err
			}
			opts.maxRestarts = int(v)
		case "--backoff-ms":
			v, err := numericFlag("--backoff-ms", 1)
			if err != nil {
				return opts, err
			}
			opts.backoffMs = v
		default:
			return opts, fmt.Errorf("unknown flag %s", args[i])
		}
		i++
	}
	return opts, fmt.Errorf("no command after --")
}

func main() {
	args := os.Args[1:]
	if len(args) == 0 {
		usage()
	}
	sub, rest := args[0], args[1:]
	switch sub {
	case "run":
		if len(rest) < 2 || rest[0] != "--" {
			usage()
		}
		os.Exit(runOnce(rest[1:]))
	case "supervise":
		opts, err := parseSupervise(rest)
		if err != nil {
			fmt.Fprintf(os.Stderr, "pmon: error: %v\n", err)
			usage()
		}
		os.Exit(supervise(opts))
	default:
		usage()
	}
}

File: examples/12-signals-and-safe-handling/cpp/src/main.cpp
// pmon v1 — a mini process supervisor (chapter 12: signals and safe handling).
//
//   pmon run -- CMD [ARGS...]
//   pmon supervise [--max-restarts N] [--backoff-ms B] -- CMD [ARGS...]
//
// The C++ take on async-signal-safe design: install NO signal handlers at
// all. SIGTERM/SIGINT/SIGHUP/SIGCHLD are blocked with sigprocmask(2) and
// converted into ordinary file-descriptor reads via signalfd(2), consumed
// from a plain poll(2) loop. No code ever runs in signal context, so nothing
// has to be async-signal-safe — except the tiny window in the forked child
// between fork() and execvp(), which restores an empty signal mask (a
// blocked, inherited SIGTERM would survive exec and make the supervised
// child unkillable).

#include <cerrno>
#include <chrono>
#include <csignal>
#include <cstdio>
#include <cstdlib>
#include <expected>
#include <print>
#include <string>
#include <string_view>
#include <system_error>
#include <utility>
#include <vector>

#include <poll.h>
#include <sys/signalfd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

namespace {

using namespace std::chrono;

class UniqueFd {
  public:
    UniqueFd() = default;
    explicit UniqueFd(int fd) : fd_(fd) {}
    UniqueFd(const UniqueFd&) = delete;
    UniqueFd& operator=(const UniqueFd&) = delete;
    UniqueFd(UniqueFd&& other) noexcept : fd_(std::exchange(other.fd_, -1)) {}
    UniqueFd& operator=(UniqueFd&& other) noexcept {
        if (this != &other) {
            reset();
            fd_ = std::exchange(other.fd_, -1);
        }
        return *this;
    }
    ~UniqueFd() { reset(); }
    [[nodiscard]] int get() const { return fd_; }
    [[nodiscard]] bool valid() const { return fd_ >= 0; }

  private:
    void reset() {
        if (fd_ >= 0) {
            ::close(fd_);
            fd_ = -1;
        }
    }
    int fd_ = -1;
};

[[nodiscard]] std::error_code errno_ec() {
    return {errno, std::generic_category()};
}

[[noreturn]] void die(std::string_view what, std::error_code ec) {
    std::println(stderr, "pmon: error: {}: {}", what, ec.message());
    std::exit(1);
}

[[noreturn]] void usage() {
    std::println(stderr, "usage: pmon run -- CMD [ARGS...]");
    std::println(stderr,
                 "       pmon supervise [--max-restarts N] [--backoff-ms B] "
                 "-- CMD [ARGS...]");
    std::exit(2);
}

// Fork + exec CMD. The child restores an empty signal mask before execvp so
// it does not inherit the parent's blocked TERM/INT/HUP/CHLD set; after
// fork() only async-signal-safe calls happen (sigprocmask, execvp, write,
// _exit).
[[nodiscard]] std::expected<pid_t, std::error_code>
spawn(const std::vector<std::string>& cmd) {
    std::vector<char*> argv;
    argv.reserve(cmd.size() + 1);
    for (const auto& arg : cmd) {
        argv.push_back(const_cast<char*>(arg.c_str()));
    }
    argv.push_back(nullptr);

    const pid_t pid = ::fork();
    if (pid < 0) {
        return std::unexpected(errno_ec());
    }
    if (pid == 0) {
        sigset_t none;
        sigemptyset(&none);
        ::sigprocmask(SIG_SETMASK, &none, nullptr);
        ::execvp(argv[0], argv.data());
        constexpr char msg[] = "pmon: error: exec failed\n";
        [[maybe_unused]] auto n = ::write(STDERR_FILENO, msg, sizeof msg - 1);
        ::_exit(127);
    }
    std::println("pmon: started pid {}", pid);
    return pid;
}

struct ChildExit {
    int status = 0;  // exit status when !signaled
    int signo = 0;   // terminating signal when signaled
    bool signaled = false;
};

[[nodiscard]] ChildExit decode(int wstatus) {
    if (WIFSIGNALED(wstatus)) {
        return {.status = 0, .signo = WTERMSIG(wstatus), .signaled = true};
    }
    return {.status = WEXITSTATUS(wstatus), .signo = 0, .signaled = false};
}

void report(pid_t pid, const ChildExit& ce) {
    if (ce.signaled) {
        std::println("pmon: child {} killed signal={}", pid, ce.signo);
    } else {
        std::println("pmon: child {} exited status={}", pid, ce.status);
    }
}

// Blocking reap that tolerates EINTR (cannot fire while the supervisor's
// mask blocks everything we watch, but `run` reaps with the default mask).
[[nodiscard]] std::expected<ChildExit, std::error_code> reap_blocking(pid_t pid) {
    int wstatus = 0;
    for (;;) {
        if (::waitpid(pid, &wstatus, 0) >= 0) {
            return decode(wstatus);
        }
        if (errno != EINTR) {
            return std::unexpected(errno_ec());
        }
    }
}

int run_once(const std::vector<std::string>& cmd) {
    const auto pid = spawn(cmd);
    if (!pid) {
        die("fork", pid.error());
    }
    const auto ce = reap_blocking(*pid);
    if (!ce) {
        die("waitpid", ce.error());
    }
    report(*pid, *ce);
    return ce->signaled ? 128 + ce->signo : ce->status;
}

struct SuperviseOpts {
    int max_restarts = 5;
    long backoff_ms = 100;
    std::vector<std::string> cmd;
};

int supervise(const SuperviseOpts& opts) {
    // Block the signals we care about, then receive them as fd reads.
    sigset_t mask;
    sigemptyset(&mask);
    for (const int signo : {SIGTERM, SIGINT, SIGHUP, SIGCHLD}) {
        sigaddset(&mask, signo);
    }
    if (::sigprocmask(SIG_BLOCK, &mask, nullptr) < 0) {
        die("sigprocmask", errno_ec());
    }
    UniqueFd sfd(::signalfd(-1, &mask, SFD_CLOEXEC));
    if (!sfd.valid()) {
        die("signalfd", errno_ec());
    }

    auto start = [&] {
        const auto pid = spawn(opts.cmd);
        if (!pid) {
            die("fork", pid.error());
        }
        return *pid;
    };
    // SIGTERM the child and reap it. Used by shutdown and reload paths.
    auto stop_child = [&](pid_t pid) {
        if (::kill(pid, SIGTERM) == 0 || errno == ESRCH) {
            if (const auto ce = reap_blocking(pid); !ce) {
                die("waitpid", ce.error());
            }
        }
    };

    pid_t child = start();
    bool running = true;  // false: no child alive, waiting out a backoff
    int restarts = 0;
    long backoff = opts.backoff_ms;
    steady_clock::time_point deadline{};

    for (;;) {
        int timeout = -1;
        if (!running) {
            const auto left = deadline - steady_clock::now();
            timeout =
                left <= 0ns
                    ? 0
                    : static_cast<int>(
                          duration_cast<milliseconds>(left).count()) +
                          1;
        }
        pollfd pfd{.fd = sfd.get(), .events = POLLIN, .revents = 0};
        const int nready = ::poll(&pfd, 1, timeout);
        if (nready < 0) {
            if (errno == EINTR) {
                continue;
            }
            die("poll", errno_ec());
        }
        if (nready == 0) {
            // Backoff elapsed: bring the child back.
            child = start();
            running = true;
            continue;
        }

        signalfd_siginfo si{};
        const ssize_t nread = ::read(sfd.get(), &si, sizeof si);
        if (nread != static_cast<ssize_t>(sizeof si)) {
            die("read signalfd", errno_ec());
        }

        switch (si.ssi_signo) {
        case SIGTERM:
        case SIGINT: {
            if (running) {
                stop_child(child);
            }
            std::println("pmon: shutting down ({})",
                         si.ssi_signo == SIGTERM ? "SIGTERM" : "SIGINT");
            return 0;
        }
        case SIGHUP: {
            std::println("pmon: reload requested");
            if (running) {
                stop_child(child);
            }
            restarts = 0;
            backoff = opts.backoff_ms;
            child = start();
            running = true;
            break;
        }
        case SIGCHLD: {
            int wstatus = 0;
            const pid_t reaped = ::waitpid(child, &wstatus, WNOHANG);
            if (!running || reaped != child) {
                break;  // stale notification (e.g. already reaped on reload)
            }
            const ChildExit ce = decode(wstatus);
            report(child, ce);
            if (!ce.signaled && ce.status == 0) {
                return 0;
            }
            if (restarts >= opts.max_restarts) {
                std::println("pmon: giving up after {} restarts",
                             opts.max_restarts);
                return 1;
            }
            ++restarts;
            std::println("pmon: restart #{} (backoff {}ms)", restarts, backoff);
            deadline = steady_clock::now() + milliseconds(backoff);
            backoff *= 2;
            running = false;
            break;
        }
        default:
            break;
        }
    }
}

// Parse "[flags] -- CMD [ARGS...]" for supervise.
[[nodiscard]] std::expected<SuperviseOpts, std::string>
parse_supervise(const std::vector<std::string>& args) {
    SuperviseOpts opts;
    std::size_t i = 0;
    auto numeric_flag = [&](const char* name, long min,
                            long& out) -> std::expected<void, std::string> {
        if (i + 1 >= args.size()) {
            return std::unexpected(std::string(name) + " needs a value");
        }
        const std::string& value = args[++i];
        char* end = nullptr;
        errno = 0;
        const long parsed = std::strtol(value.c_str(), &end, 10);
        if (errno != 0 || end == value.c_str() || *end != '\0' ||
            parsed < min) {
            return std::unexpected("bad value for " + std::string(name));
        }
        out = parsed;
        return {};
    };
    while (i < args.size()) {
        const std::string& arg = args[i];
        if (arg == "--") {
            ++i;
            break;
        }
        if (arg == "--max-restarts") {
            long v = 0;
            if (auto r = numeric_flag("--max-restarts", 0, v); !r) {
                return std::unexpected(r.error());
            }
            opts.max_restarts = static_cast<int>(v);
        } else if (arg == "--backoff-ms") {
            long v = 0;
            if (auto r = numeric_flag("--backoff-ms", 1, v); !r) {
                return std::unexpected(r.error());
            }
            opts.backoff_ms = v;
        } else {
            return std::unexpected("unknown flag " + arg);
        }
        ++i;
    }
    for (; i < args.size(); ++i) {
        opts.cmd.push_back(args[i]);
    }
    if (opts.cmd.empty()) {
        return std::unexpected("no command after --");
    }
    return opts;
}

}  // namespace

int main(int argc, char** argv) {
    std::setvbuf(stdout, nullptr, _IOLBF, 0);  // line-buffered even into a file

    const std::vector<std::string> args(argv + 1, argv + argc);
    if (args.empty()) {
        usage();
    }
    const std::string& sub = args[0];
    const std::vector<std::string> rest(args.begin() + 1, args.end());

    if (sub == "run") {
        if (rest.size() < 2 || rest[0] != "--") {
            usage();
        }
        return run_once({rest.begin() + 1, rest.end()});
    }
    if (sub == "supervise") {
        const auto opts = parse_supervise(rest);
        if (!opts) {
            std::println(stderr, "pmon: error: {}", opts.error());
            usage();
        }
        return supervise(*opts);
    }
    usage();
}
